apps/api/src/care-activity/care-activity.service.ts

```diff
@@ -190,54 +222,74 @@ export class CareActivityService {
       });
     }
 
-    // deconstruct
-    const { bundle: bundleId, careLocations: careLocationIds, ...careActivityLiterals } = data;
+    if (!careActivity.careLocations.some(u => u.id === unitId)) {
+      throw new NotFoundException({
+        message: 'Cannot update care activity: not related to the unit',
+        data: { id, unitId: data.unitId },
+      });
+    }
 
     // if bundle is updated, fetch and update entity
-    if (bundleId) {
+    if (bundleId && careActivity.bundle.id !== bundleId) {
       const bundle = await this.bundleRepo.findOneBy({ id: bundleId });
       if (!bundle) {
         throw new NotFoundException({
           message: 'Cannot update care activity: Bundle not found',
           data: { id: bundleId },
         });
       }
-
-      careActivity.bundle = bundle;
     }
 
-    // if care settings / units / care locations are updated, fetch and update entities
-    if (Array.isArray(careLocationIds)) {
-      const careLocations = await this.unitService.getManyByIds(careLocationIds);
+    await this.careActivityRepo.manager.transaction(async manager => {
+      await manager.update(CareActivity, id, {
+        bundle: { id: bundleId },
+        displayName: name,
+        description,
+        activityType,
+      });
+
+      // update allowed activities for the unit
+      const currrentAllowedActivityMap = new Map(
+        careActivity.allowedActivities.map(aa => [aa.id, aa]),
+      );
+      const allowedActivitiesToDelete = allowedActivities
+        .filter(aa => aa.id && aa.permission === 'N')
+        .map(aa => aa.id);
```

Comment:
I would update the variable names to be more descriptive or provide a comment to describe the purpose of this function.

Reply:
added a comment